### PR TITLE
Finished the MVP of the VMM

### LINKED ISSUE #11: Create a virtual memory manager for the x86_64 ISA
Create a virtual memory manager for the x86_64 ISA that can do the following:

- [x] ~~Enable the use of PCIDs~~*
- [x] Create a page map
- [x] ~~Implement `Clone` and `Drop` for page map~~*
- [x] Load a page map
- [x] Map normal 4 KiB pages
- [x] Unmap normal pages
- [x] Map large pages
- [x] Unmap large pages
- [x] Detect CPU support for huge pages
- [x] Map huge pages
- [x] Unmap huge pages
- [x] ~~Handle page fault exceptions~~**
- [x] ~~Make use of the page attribute table (PAT) to to set memory attributes for different regions~~***

*Postponed until there are multiple PageMaps
**Postponed until there is a userspace
***Moved to a separate issue due to the possibility of breaking the bootloader provided PageMap

### PR DESCRIPTION
**Pull request by mdpcs3544**
This PR closes #11 and completes all of the non-stricken out tasks listed therein.

### CODE REVIEW

File: charlotte_core/src/arch/x86_64/mod.rs [hunk 1]
```diff
@@ -231,4 +270,119 @@ impl Api {
...
+            Ok(vaddr) => vaddr,
+            Err(e) => {
+                panic!("Failed to create VirtualAddress: {:?}", e);
+            }
+        };
+        pm.map_huge_page(
+            vaddr,
+            huge_frame,
+            PteFlags::Write as u64 | PteFlags::Global as u64 | PteFlags::NoExecute as u64,
+        );
```
Comment (Wrench56) [inline charlotte_core/src/arch/x86_64/mod.rs:402]: Please ignore or include `let _ = ` to suppress the returned error. Ideally, as this is testing the VMM, handeling the error and writing out a log accordingly would be even better.

File: charlotte_core/src/arch/x86_64/mod.rs [hunk 2]
```diff
@@ -231,4 +270,119 @@ impl Api {
...
+            Ok(vaddr) => vaddr,
+            Err(e) => {
+                panic!("Failed to create VirtualAddress: {:?}", e);
+            }
+        };
+        pm.map_large_page(
+            vaddr,
+            large_frame,
+            PteFlags::Write as u64 | PteFlags::Global as u64 | PteFlags::NoExecute as u64,
+        );
```
Comment (Wrench56) [inline charlotte_core/src/arch/x86_64/mod.rs:369]: Please ignore or include `let _ =`  to suppress the returned error. Ideally, as this is testing the VMM, handeling the error and writing out a log accordingly would be even better.

File: charlotte_core/src/arch/x86_64/mod.rs [hunk 3]
```diff
@@ -231,4 +270,119 @@ impl Api {
...
+            Ok(vaddr) => vaddr,
+            Err(e) => {
+                panic!("Failed to create VirtualAddress: {:?}", e);
+            }
+        };
+        pm.map_page(
+            vaddr,
+            frame,
+            PteFlags::Write as u64 | PteFlags::Global as u64 | PteFlags::NoExecute as u64,
+        );
```
Comment (Wrench56) [inline charlotte_core/src/arch/x86_64/mod.rs:338]: Please ignore or include `let _ =`  to suppress the returned error. Ideally, as this is testing the VMM, handeling the error and writing out a log accordingly would be even better.

File: charlotte_core/src/arch/x86_64/memory/mod.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+use crate::memory::pmm::Error as PmmError;
+use spin::lazy::Lazy;
+
+/// The number of significant binary digits in a physical address
+pub static PADDR_SIGBITS: Lazy<u8> = Lazy::new(|| {
+    let cpuid_result = unsafe { __cpuid_count(0x80000008, 0) };
+    (cpuid_result.eax & 0xFF) as u8
+});
+
+/// This is the number of significant binary digits in a virtual (linear) address
```
Comment (Wrench56) [inline charlotte_core/src/arch/x86_64/memory/mod.rs:15]: Please use the `#[allow(unused)]` directive to ignore warnings in our DevOps. You can lint your code manually by running `./tools/lint.sh`. Other places have bits of code that is unused for now, ignoring them would be great. Disabling this lint is another solution. Either way, we should keep our lints clean of warnings.

**Review by Wrench56 — COMMENTED** [review]
Added some comments about pedantic linting and formatting.

**Review by parkerstafford — APPROVED** [review]

File: charlotte_core/src/arch/x86_64/mod.rs (hunk 3 above)
Comment (GranitBebok) [reply charlotte_core/src/arch/x86_64/mod.rs:338]: Agree with wrench in this one.

File: charlotte_core/src/arch/x86_64/mod.rs (hunk 2 above)
Comment (GranitBebok) [reply charlotte_core/src/arch/x86_64/mod.rs:369]: +1

File: charlotte_core/src/arch/x86_64/mod.rs (hunk 1 above)
Comment (GranitBebok) [reply charlotte_core/src/arch/x86_64/mod.rs:402]: +1

**Review by xadaemon — CHANGES_REQUESTED** [review]
So far on a lint pass I agree with wrench

**Review by xadaemon — APPROVED** [review]

**Review by Wrench56 — APPROVED** [review]
LGTM!

**Pull request merged**